File: ios/Sources/PrinterPlugin/PrinterPlugin.swift
import Foundation
import Capacitor

/// Capacitor bridge for the Printer plugin. The print logic lives in
/// `Printer.swift`; this class only maps plugin calls onto it.
@objc(PrinterPlugin)
public class PrinterPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "PrinterPlugin"
    public let jsName = "Printer"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "printWebView", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = Printer()

    @objc func printWebView(_ call: CAPPluginCall) {
        let jobName = implementation.resolveJobName(call.getString("name"))

        DispatchQueue.main.async { [weak self] in
            guard let self = self, let webView = self.bridge?.webView else {
                call.reject("Web view is not available", "WEBVIEW_UNAVAILABLE")
                return
            }

            self.implementation.printWebView(
                webView,
                anchoredTo: self.bridge?.viewController?.view,
                jobName: jobName
            ) { result in
                switch result {
                case .success:
                    call.resolve()
                case .failure(let error):
                    call.reject(error.localizedDescription, "PRINT_FAILED", error)
                }
            }
        }
    }
}

File: android/src/main/java/co/baseplay/capacitor/printer/PrinterPlugin.java
package co.baseplay.capacitor.printer;

import android.webkit.WebView;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * Capacitor bridge for the Printer plugin. The print logic lives in
 * {@link Printer}; this class only maps plugin calls onto it.
 */
@CapacitorPlugin(name = "Printer")
public class PrinterPlugin extends Plugin {

    private final Printer implementation = new Printer();

    @PluginMethod
    public void printWebView(PluginCall call) {
        final String jobName = implementation.resolveJobName(call.getString("name"));

        getActivity().runOnUiThread(() -> {
            WebView webView = getBridge().getWebView();
            if (webView == null) {
                call.reject("Web view is not available", "WEBVIEW_UNAVAILABLE");
                return;
            }

            try {
                implementation.printWebView(getContext(), webView, jobName);
                call.resolve();
            } catch (Exception e) {
                call.reject(e.getMessage(), "PRINT_FAILED", e);
            }
        });
    }
}
